### PR TITLE
Add support for IPyWidgets 8

### LINKED ISSUE #8552: Add support for ipywidgets 8
Packages such as `tqdm` will not work with the latest version of IPyWidgets 8
* [ ] [Explore supporting IPyWidgets 8](https://github.com/microsoft/vscode-jupyter/issues/11217)

# Work around - Please install ipywidgets==7.7.2

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #8552

### CODE REVIEW

File: src/kernels/kernelWorkingFolder.node.ts [hunk 1]
```diff
@@ -36,7 +36,11 @@ export class KernelWorkingFolder {
         let suggestedDir = await this.calculateWorkingDirectory(kernel.resourceUri);
         if (suggestedDir && (await this.fs.exists(suggestedDir))) {
             return suggestedDir;
-        } else if (kernel.resourceUri && (await this.fs.exists(kernel.resourceUri))) {
+        } else if (
+            kernel.resourceUri &&
+            kernel.resourceUri.scheme !== 'untitled' &&
```
Comment (DonJayamanne) [inline src/kernels/kernelWorkingFolder.node.ts:41]: Debt = we weren't checking if it was untitled, 
No point chekcing if untitled files exist on disc or not, as the answer is obvious

File: src/kernels/kernelProvider.base.ts [hunk 2]
```diff
@@ -137,6 +138,7 @@ export abstract class BaseCoreKernelProvider implements IKernelProvider {
                 `Disposing kernel associated with ${getDisplayPath(notebook.uri)}, isClosed=${notebook.isClosed}`
             );
             this.pendingDisposables.add(kernelToDispose.kernel);
+            this.executions.get(kernelToDispose.kernel)?.dispose();
```
Comment (DonJayamanne) [inline src/kernels/kernelProvider.base.ts:141]: debt

File: src/kernels/kernelExecution.ts [hunk 3]
```diff
@@ -71,6 +72,9 @@ export class NotebookKernelExecution implements INotebookKernelExecution {
         kernel.addHook('willRestart', (sessionPromise) => this.onWillRestart(sessionPromise), this, this.disposables);
         this.disposables.push(this._onPreExecute);
     }
+    public dispose() {
+        disposeAllDisposables(this.disposables);
```
Comment (DonJayamanne) [inline src/kernels/kernelExecution.ts:76]: debt

File: src/kernels/types.ts [hunk 4]
```diff
@@ -391,7 +393,7 @@ export interface IKernel extends IBaseKernel {
     readonly creator: 'jupyterExtension';
 }
 
-export interface INotebookKernelExecution {
+export interface INotebookKernelExecution extends IDisposable {
```
Comment (DonJayamanne) [inline src/kernels/types.ts:396]: debt

File: src/notebooks/controllers/vscodeNotebookController.ts [hunk 5]
```diff
@@ -470,44 +470,7 @@ export class VSCodeNotebookController implements Disposable, IVSCodeNotebookCont
...
-                    '@vscode',
-                    'jupyter-ipywidgets7',
-                    'dist',
-                    'ipywidgets.js'
-                )
-            );
         }
-        scripts.push(
-            ...[
-                Uri.joinPath(
```
Comment (DonJayamanne) [inline src/notebooks/controllers/vscodeNotebookController.ts:501]: All of these are now statically defined in package.json

File: .github/workflows/build-test.yml [hunk 6]
```diff
@@ -287,47 +288,62 @@ jobs:
...
             tags: '^[^@]+$|@mandatory'
+            os: windows-latest
+            ipywidgetsVersion: ''
+        # IPyWidgets 8
+          - jupyterConnection: raw
+            python: python
+            pythonVersion: '3.10'
+            tags: '@widgets'
+            os: ubuntu-latest
+            ipywidgetsVersion: '8'
```
Comment (DonJayamanne) [inline .github/workflows/build-test.yml:312]: Added Ci tgests for IPyWidgets 8
We need to have tests for both 7 & 8

File: build/venv-test-ipywidgets8-requirements.txt [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+# List of requirements for ipython tests
+numpy
+pandas
+# Install jupyter itself so we end up with a kernel
+jupyter
+# List of requirements for conda environments that cannot be installed using conda
+livelossplot
+versioneer
+pythreejs
+ipywidgets
```
Comment (DonJayamanne) [inline build/venv-test-ipywidgets8-requirements.txt:15]: This is a copy of `venv-test-requirements.txt` with the difference of having ipywidgets not pinned to a specific version

File: build/webpack/webpack.datascience-ui.config.builder.js [hunk 8]
```diff
@@ -121,12 +121,60 @@ function buildConfiguration(bundle) {
...
+                constants.ExtensionRootDir,
+                'out',
+                'node_modules',
+                '@vscode',
+                'jupyter-ipywidgets7',
+                'dist',
+                'ipywidgets.js'
+            )
+        });
+        filesToCopy.push({
```
Comment (amunger) [inline build/webpack/webpack.datascience-ui.config.builder.js:145]: I assume this is supposed to be in an `else` block?

Comment (DonJayamanne) [reply build/webpack/webpack.datascience-ui.config.builder.js:145]: no, its still here, 
just made it separate to highlight the two sets, one for 7 and one for 8

File: src/kernels/kernel.ts [hunk 9]
```diff
@@ -53,13 +58,41 @@ import { Cancellation, isCancellationError } from '../platform/common/cancellati
 import { KernelProgressReporter } from '../platform/progress/kernelProgressReporter';
 import { DisplayOptions } from './displayOptions';
 import { SilentExecutionErrorOptions } from './helpers';
+import dedent from 'dedent';
+import { IAnyMessageArgs } from '@jupyterlab/services/lib/kernel/kernel';
+
+const widgetVersionOutPrefix = 'e976ee50-99ed-4aba-9b6b-9dcd5634d07d:IPyWidgets:';
```
Comment (amunger) [inline src/kernels/kernel.ts:64]: what is this value? where is it from?

Comment (DonJayamanne) [reply src/kernels/kernel.ts:64]: its just a GUUID, you can see how it is used, in print statements.
see here (further below)
```typescript
        try:
            import ipywidgets as _VSCODE_ipywidgets
            print("${widgetVersionOutPrefix}" + _VSCODE_ipywidgets.__version__)
            del _VSCODE_ipywidgets
        except:
            pass
        `;

            const version = await this.executeSilently(session, [codeToDetermineIPyWidgetsVersion]).catch((ex) =>
                traceError('Failed to determine version of IPyWidgets', ex)
            );
            traceError('Determined IPyKernel Version', JSON.stringify(version));
            if (Array.isArray(version)) {
                const isVersion8 = version.some((output) =>
                    (output.text || '')?.toString().includes(`${widgetVersionOutPrefix}8.`)
                );
                const isVersion7 = version.some((output) =>
                    (output.text || '')?.toString().includes(`${widgetVersionOutPrefix}7.`)
                );
```

**Review by amunger — APPROVED** [review]

**Pull request merged**